File: 04-april/07-walking-robot-simulation-ii/solution.js
/**
 * @param {number} width
 * @param {number} height
 */
var Robot = function(width, height) {
    this.width = width;
    this.height = height;
    this.perimeter = 2 * (width + height) - 4;
    this.index = 0;
    this.moved = false;
};

/**
 * @param {number} num
 * @return {void}
 */
Robot.prototype.step = function(num) {
    this.moved = true;
    let steps = num % this.perimeter;
    if (steps === 0) {
        steps = this.perimeter;
    }
    this.index = (this.index + steps) % this.perimeter;
};

/**
 * @return {number[]}
 */
Robot.prototype.getPos = function() {
    return [this.getX(), this.getY()];
};

/**
 * @return {string}
 */
Robot.prototype.getDir = function() {
    if (this.index === 0) {
        return this.moved ? "South" : "East";
    }
    if (this.index < this.width) {
        return "East";
    }
    if (this.index < this.width + this.height - 1) {
        return "North";
    }
    if (this.index < 2 * this.width + this.height - 2) {
        return "West";
    }
    return "South";
};

Robot.prototype.getX = function() {
    if (this.index < this.width) {
        return this.index;
    }
    if (this.index < this.width + this.height - 1) {
        return this.width - 1;
    }
    if (this.index < 2 * this.width + this.height - 2) {
        return 2 * this.width + this.height - 3 - this.index;
    }
    return 0;
};

Robot.prototype.getY = function() {
    if (this.index < this.width) {
        return 0;
    }
    if (this.index < this.width + this.height - 1) {
        return this.index - this.width + 1;
    }
    if (this.index < 2 * this.width + this.height - 2) {
        return this.height - 1;
    }
    return this.perimeter - this.index;
};

/**
 * Your Robot object will be instantiated and called as such:
 * var obj = new Robot(width, height)
 * obj.step(num)
 * var param_2 = obj.getPos()
 * var param_3 = obj.getDir()
 */

File: 04-april/07-walking-robot-simulation-ii/solution.java
class Robot {
    private final int width;
    private final int height;
    private final int perimeter;
    private int index;
    private boolean moved;

    public Robot(int width, int height) {
        this.width = width;
        this.height = height;
        this.perimeter = 2 * (width + height) - 4;
        this.index = 0;
        this.moved = false;
    }

    public void step(int num) {
        moved = true;
        int steps = num % perimeter;
        if (steps == 0) {
            steps = perimeter;
        }
        index = (index + steps) % perimeter;
    }

    public int[] getPos() {
        return new int[] {getX(), getY()};
    }

    public String getDir() {
        if (index == 0) {
            return moved ? "South" : "East";
        }
        if (index < width) {
            return "East";
        }
        if (index < width + height - 1) {
            return "North";
        }
        if (index < 2 * width + height - 2) {
            return "West";
        }
        return "South";
    }

    private int getX() {
        if (index < width) {
            return index;
        }
        if (index < width + height - 1) {
            return width - 1;
        }
        if (index < 2 * width + height - 2) {
            return 2 * width + height - 3 - index;
        }
        return 0;
    }

    private int getY() {
        if (index < width) {
            return 0;
        }
        if (index < width + height - 1) {
            return index - width + 1;
        }
        if (index < 2 * width + height - 2) {
            return height - 1;
        }
        return perimeter - index;
    }
}
